### ISSUE TITLE
optimize align for scalars at least

### ISSUE THREAD

**Issue by dcherian** [pydata/xarray#8350]
### What happened?

Here's a simple rescaling calculation:
```python
import numpy as np
import xarray as xr

ds = xr.Dataset(
    {"a": (("x", "y"), np.ones((300, 400))), "b": (("x", "y"), np.ones((300, 400)))}
)
mean = ds.mean() # scalar
std = ds.std() # scalar
rescaled = (ds - mean) / std
```

The profile for the last line shows 30% (!!!) time spent in `align` (really `reindex_like`) except there's nothing to reindex when only scalars are involved! 
<img width="1088" alt="image" src="https://github.com/pydata/xarray/assets/2448579/426de01a-e05c-46cf-9f86-25333b4843be">

This is a small example inspired by a ML pipeline where this normalization is happening very many times in a tight loop.

cc @benbovy 

### What did you expect to happen?

A fast path for when no reindexing needs to happen.

### MVCE confirmation

- [ ] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [ ] Complete example — the example is self-contained, including all data and the text of any traceback.
- [ ] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [ ] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [ ] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Environment

<details>

</details>

**Comment by benbovy** [comment]
Maybe it would be enough to refactor `Aligner.reindex_all()` like this:

```python
    def reindex_all(self) -> None:
        results = []

        for obj, matching_indexes in zip(
            self.objects, self.objects_matching_indexes
        ):
            if any(self.reindex[k] for k in matching_indexes):
                results.append(self._reindex_one(obj, matching_indexes))
            else:
                results.append(obj.copy(deep=self.copy))

        self.results = tuple(results)
```

**Comment by benbovy** [comment]
> Maybe it would be enough to refactor Aligner.reindex_all() like this

It depends on the profiling results below `align` in the call stack: if most of the time is spent in `Aligner.normalize_index` or `Aligner.find_matching_*`, optimizing this case would require a bit more work. @dcherian could you paste another screenshot with the root set at that level, please?

**Comment by dcherian** [comment]
Does this help?
<img width="1349" alt="image" src="https://github.com/pydata/xarray/assets/2448579/b1fc7db9-ada5-4387-9765-5e26155aacae">

**Comment by benbovy** [comment]
Yes, thanks. `dataset.copy()` is 50% of the align time. Looks like the code snippet in my previous comment won't improve the performance issue much or at all? Should we rather look into if/how to optimize `copy` instead?

**Comment by benbovy** [comment]
I guess that in this case we could skip copying objects that don't need reindexing, but is it always safe or desirable? Do we have an option for this (or skip alignment altogether)?
